osbuild.py

```diff
@@ -201,27 +196,51 @@ def print_header(title, options, machine_name):
     print(f"\t# nsenter -a --wd=/root -t `machinectl show {machine_name} -p Leader --value`")
     print()
 
+class Pipeline:
+    def __init__(self, pipeline, content_store):
+        m = hashlib.sha256()
+        m.update(json.dumps(pipeline, sort_keys=True).encode())
+
+        self.id = m.hexdigest()
+        self.base = pipeline.get("base")
+        self.stages = pipeline.get("stages")
+        self.assembler = pipeline.get("assembler")
+        self.content_store = content_store
 
-def run(pipeline, input_dir, output_dir, interactive=False):
-    results = {
-        "stages": []
-    }
-    with BuildRoot() as buildroot, tmpfs() as tree:
-        for i, stage in enumerate(pipeline["stages"], start=1):
-            name = stage["name"]
-            options = stage.get("options", {})
-            if interactive:
-                print_header(f"{i}. {name}", options, buildroot.machine_name)
-            r = buildroot.run_stage(stage, tree, input_dir, interactive)
-            results["stages"].append(r)
-
-        assembler = pipeline.get("assembler")
-        if assembler:
-            name = assembler["name"]
-            options = assembler.get("options", {})
-            if interactive:
-                print_header(f"Assembling: {name}", options, buildroot.machine_name)
-            r = buildroot.run_assembler(assembler, tree, input_dir, output_dir, interactive)
-            results["assembler"] = r
-
-    return results
+        os.makedirs(content_store, exist_ok=True)
+
+    def run(self, output_dir, interactive=False):
+        results = {
+            "stages": []
+        }
+        with BuildRoot() as buildroot, tmpfs() as tree:
+            if self.base:
+                input_tree = os.path.join(self.content_store, self.base)
+
+                subprocess.run(["cp", "-a", f"{input_tree}/.", tree], check=True)
+
+            if self.stages:
```

Comment:
See above. No need to check for this (and we save a level of indentation) if `self.stages` is always a list.